### PR TITLE
Migrate session functionality from Django

### PR DESCRIPTION
**Pull request by blazebissar**
Checkout the [Engineer Workflow Doc](https://docs.google.com/presentation/d/1pdzuHG-yO7_qW6jiByaGTLztKZ6b72lBNqSdNT5WZQ8/edit) to make sure you're contributing correctly! :)

**[Tickets](https://trello.com/b/dYOQKJGv/reclass)** 
Trello Ticket: https://trello.com/c/OzyFBuVh/404-django-step-2-4-implement-toynetsession-rest-api-resource
Additional Notes: Steps 2-4 complete as well as step 5 from https://trello.com/c/eufeDc9q/405-django-step-5-test-toynetsession-and-toynetsessionbyid

**[Dependencies](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.czpk4cbi7z4g)**
None at this time; though `users` table development will potentially affect some relationships

**[Explain Impact on DB](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.285lgcs4z84o)**

Added dummy user data into the DB, also fixed an `INSERT` statement for the `toynet_sessions` table to use a `TEXT` `user_id`

**[Testing](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.rs692kr58cf8)**

How did you test this? List of actions & expected behavior with helpful screenshots for reviewer:

Ad hoc testing at this time until step 5 (unit tests). Validated success and failure cases for each db call as well as persistence after successful `PUT` and `PUSH` calls.

If this change is performance related, please document what tests were / should be done to verify improvement:
None.

### CODE REVIEW

File: flasksrc/session.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+class ToyNetSessionByIdGetResponseSchema(Schema):
+    topo_id = fields.Int()
+    topology = fields.Str()
+    user_id = fields.Str()
+
+class ToyNetSessionByIdPutRequestSchema(Schema):
+    toynet_session_id = fields.Int()
+
+class ToyNetSessionByIdPutResponseSchema(Schema):
+    pass
```
Comment (takakonishimura) [inline flasksrc/session.py:31]: So far I have only defined Schemas where arguments need to be parsed. I'm open to defining outputs, in which case we should create a work item on trello to apply it to the rest of the code base.

File: flasksrc/__init__.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import os
-
 from flask import Flask
 from flask_restful import Resource, Api
 from flask_jwt_extended import JWTManager
+from flask_apispec import FlaskApiSpec
 
 
 class HelloReclass(Resource):
```
Comment (scottjr632) [inline flasksrc/__init__.py:8]: Should we create a better name for this or have a better response if this is a status?

File: flasksrc/session.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+
+# Schema definitions
+class ToyNetSessionPostReq(Schema):
+    toynet_topo_id = fields.Int(required=True)
+    toynet_user_id = fields.Str(required=True)
+
+
+class ToyNetSessionPostResp(Schema):
+    status = fields.Bool(required=True)
```
Comment (scottjr632) [inline flasksrc/session.py:16]: Why do we need status?

File: flasksrc/session.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                    ' VALUES(?,?,?)',
+                    (str(toynet_topo_id), topo_rows[0]['topology'], toynet_user_id,)
+                    )
+            db.commit()
+        except Exception as e:
+            print(e.args[0])
+            abort(500, message='Failed to create new session')
+
+        return {
+            'status': True,
```
Comment (ajlacey) [inline flasksrc/session.py:80]: 'status' not used by front end

File: flasksrc/session.py (hunk 3 above)
Comment (blazebissar) [inline flasksrc/session.py:16]: `status` not needed

File: flasksrc/__init__.py (hunk 2 above)
Comment (ajlacey) [inline flasksrc/__init__.py:8]: address the #return Hello Reclass

**Comment by blazebissar** [comment]
@ajlacey we should also add in some comments and docstrings when we lint. I still need to add in the swagger information to the `readme`.

File: flasksrc/session.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    @marshal_with(ToyNetSessionByIdGetResp)
+    def get(self, toynet_session_id):
+        db = get_db()
+
+        try:
+            rows = db.execute(
+                'SELECT topo_id, topology, user_id' +#, create_time, update_time' +
+                ' FROM toynet_sessions' +
+                ' WHERE session_id = (?)',
+                (str(toynet_session_id),)
```
Comment (scottjr632) [inline flasksrc/session.py:95]: Is there a chance for SQL injection with this approach?

File: flasksrc/user.py [hunk 6]
```diff
@@ -1,33 +1,42 @@
...
+    username = fields.Str(required=True)
+
+
+class ToyNetUserLoginPostReq(Schema):
+    username = fields.Str(required=True)
+    password = fields.Str(required=True)
+
+
+class ToyNetUserLoginPostResp(Schema):
+    verified = fields.Bool(required=True)
```
Comment (scottjr632) [inline flasksrc/user.py:31]: What is the purpose of verified here?

File: flasksrc/session.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class ToyNetSessionById(MethodResource, Resource):
+    @marshal_with(ToyNetSessionByIdGetResp)
+    def get(self, toynet_session_id):
+        db = get_db()
+
+        try:
+            rows = db.execute(
+                'SELECT topo_id, topology, user_id' +#, create_time, update_time' +
```
Comment (takakonishimura) [inline flasksrc/session.py:92]: clean up comment

File: flasksrc/__init__.py [hunk 8]
```diff
@@ -50,4 +50,14 @@ def create_app(test_config=None):
     from .survey import ToyNetSurveyById
     api.add_resource(ToyNetSurveyById, '/api/survey/<string:survey_id>')
 
+    from .session import ToyNetSession
+    api.add_resource(ToyNetSession, '/api/toynet/session')
+
+    from .session import ToyNetSessionById
+    api.add_resource(ToyNetSessionById, '/api/toynet/session/<string:toynet_session_id>')
```
Comment (takakonishimura) [inline flasksrc/__init__.py:57]: Hm, should this (an other IDs) be `int`?

File: flasksrc/value.py [hunk 9]
```diff
@@ -27,22 +42,16 @@ def get(self, value_id):
 
         inspirations = [{'organization': r['organization'], 'definition': r['quote']} for r in rows]
 
+        print(inspirations)
```
Comment (takakonishimura) [inline flasksrc/value.py:45]: remove print statement

File: tests/test_values.py [hunk 10]
```diff
@@ -28,8 +26,8 @@ def test_valueById_get(client):
     rv = client.get('/api/value/5004/inspirations')
     assert rv.status_code == 200
     rv_json = json.loads(rv.data.decode('utf-8'))
-
     assert rv_json['value'] == 'Loyalty'
+    print(rv_json)
```
Comment (takakonishimura) [inline tests/test_values.py:30]: remove print statement

**Review by takakonishimura — APPROVED** [review]

File: tests/test_sessions.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    db_fd, app.config['DATABASE'] = tempfile.mkstemp()
+
+    with app.test_client() as client:
+        with app.app_context():
+            db.init_db()
+        yield client
+
+    os.close(db_fd)
+    os.unlink(app.config['DATABASE'])
+def test_topo_id_post(client):
```
Comment (takakonishimura) [inline tests/test_sessions.py:20]: two new lines above this line

File: flasksrc/session.py (hunk 7 above)
Comment (takakonishimura) [reply flasksrc/session.py:92]: Remove all `+` to align with:
https://github.com/Project-Reclass/toynet-flask/blob/main/flasksrc/value.py#L54

File: flasksrc/session.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        except ValidationError as e:
+            abort(400, message=f'malformed request: {e.messages}')
+
+        db = get_db()
+
+        new_topo = req['new_topology']
+
+        try:
+            db.execute(
+                    'UPDATE toynet_sessions' +
```
Comment (takakonishimura) [inline flasksrc/session.py:122]: Remove all `+` to align with:
https://github.com/Project-Reclass/toynet-flask/blob/main/flasksrc/value.py#L54

File: flasksrc/session.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            abort(400, message=f'malformed request: {e.messages}')
+
+        toynet_topo_id = req['toynet_topo_id']
+        toynet_user_id = req['toynet_user_id']
+
+        db = get_db()
+
+        try:
+            topo_rows = db.execute(
+                'SELECT topology' +
```
Comment (takakonishimura) [inline flasksrc/session.py:41]: Remove all `+` to align with: https://github.com/Project-Reclass/toynet-flask/blob/main/flasksrc/value.py#L54

File: flasksrc/session.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        except Exception as e:
+            print(e.args[0])
+            abort(500, message='topo_id query failed: {}'.format(toynet_topo_id))
+
+        if not len(topo_rows):
+            abort(400, message='topo_id is invalid: {}'.format(toynet_topo_id))
+
+        try:
+            user_rows = db.execute(
+                'SELECT username' +
```
Comment (takakonishimura) [inline flasksrc/session.py:55]: Remove all `+` to align with:
https://github.com/Project-Reclass/toynet-flask/blob/main/flasksrc/value.py#L54

File: flasksrc/session.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+            print(e.args[0])
+            abort(500, message='user_id query failed: {}'.format(toynet_user_id))
+
+        if not len(user_rows):
+            abort(400, message='user_id is invalid: {}'.format(toynet_user_id))
+
+        try:
+            cur = db.cursor()
+            cur.execute(
+                    'INSERT INTO toynet_sessions(topo_id, topology, user_id)' +
```
Comment (takakonishimura) [inline flasksrc/session.py:70]: Remove all `+` to align with:
https://github.com/Project-Reclass/toynet-flask/blob/main/flasksrc/value.py#L54

File: tests/test_sessions.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    """Check the POST topo_id returns a session id if valid or error if not"""
+    client.post(
+          '/api/user',
+          data={
+              'username': 'Arthur@projectreclass.org',
```
Comment (takakonishimura) [inline tests/test_sessions.py:25]: lowercase `Arthur` to `arthur` for consistency with other test files

File: tests/test_sessions.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+              'password': 'BaLtH@$0R',
+              'first_name': 'Arthur',
+          },
+    )
+    url = "/api/toynet/session"
+
+    pv = ({'toynet_topo_id': 1, 'toynet_user_id': 'Arthur@projectreclass.org',})
```
Comment (takakonishimura) [inline tests/test_sessions.py:32]: lowercase `Arthur` to `arthur` for consistency with other test files

File: tests/test_sessions.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    rv = client.post(url, data = pv,)
```
Comment (takakonishimura) [inline tests/test_sessions.py:33]: Please put this on multiple lines with `url` and `pv` embedded to align with other test files for readability (this is just a style preference)

File: tests/test_sessions.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+              'first_name': 'Arthur',
+          },
+    )
+    url = "/api/toynet/session"
+
+    pv = ({'toynet_topo_id': 1, 'toynet_user_id': 'Arthur@projectreclass.org',})
+    rv = client.post(url, data = pv,) 
+
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    #print(rv_json['message'])
```
Comment (takakonishimura) [inline tests/test_sessions.py:36]: remove print statement

File: tests/test_sessions.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    assert rv.status_code == 201
+    assert rv_json['status'] == True
+    # check that a session id was returned
+    assert type(rv_json['toynet_session_id']) == int
+
+    #Invalid toynet_topo_id number
+    pv = ({'toynet_topo_id': -5, 'toynet_user_id': 'Arthur@projectreclass.org',})
```
Comment (takakonishimura) [inline tests/test_sessions.py:43]: here as well, please apply this style to all test files, or make this test file match the rest :)

Comment (takakonishimura) [reply tests/test_sessions.py:43]: `Arthur` -> `arthur`

File: tests/test_sessions.py [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    rv = client.post(url, data = pv,)
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    print(rv_json['message'])
```
Comment (takakonishimura) [inline tests/test_sessions.py:46]: Remove print statement

File: tests/test_sessions.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+    assert rv.status_code == 400
+    assert rv_json['message'] == 'topo_id is invalid: -5'
+
+    #invalid user id
+    pv = ({'toynet_topo_id': 1, 'toynet_user_id': "Loser@projectreclass.org",}) #this should produce an error because we have no losers. #facts
```
Comment (takakonishimura) [inline tests/test_sessions.py:51]: `Loser` -> `loser`

File: tests/test_sessions.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    rv = client.post(url, data = pv,)
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    print(rv_json['message'])
```
Comment (takakonishimura) [inline tests/test_sessions.py:54]: Remove print statement

File: tests/test_sessions.py [hunk 24]
```diff
@@ -0,0 +1,75 @@
+import os
```
Comment (takakonishimura) [inline tests/test_sessions.py:1]: Please use single quotes wherever possible to match the rest of the repository

File: tests/test_sessions.py [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    pv = ({'toynet_topo_id': 1, 'toynet_user_id': "Loser@projectreclass.org",}) #this should produce an error because we have no losers. #facts
+    rv = client.post(url, data = pv,)
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    print(rv_json['message'])
+    assert rv.status_code == 400
+    assert rv_json['message'] == 'user_id is invalid: Loser@projectreclass.org'
+
+def test_session_by_id_get(client):
+    url = "/api/toynet/session/"#session_id
+    #200 on success
```
Comment (takakonishimura) [inline tests/test_sessions.py:60]: remove comment

File: tests/test_sessions.py [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+    session_id = "1"
+    rv = client.get(url+session_id)
```
Comment (takakonishimura) [inline tests/test_sessions.py:62]: `url = '/api/toynet/session/1'`

or

```
session_id = '1'
url = '/api/toynet/session/' + session_id
```

File: tests/test_sessions.py [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+    assert rv.status_code == 200
+    rv_json = json.loads(rv.data.decode('utf-8'))
+    '''expected rv_json output
+    {
+        'topo_id': <int>,
+        'topology': <XML?>,
+        'user_id': <str>,
+    }
+    '''
+    print(rv_json)
```
Comment (takakonishimura) [inline tests/test_sessions.py:72]: Remove print statement

File: tests/test_sessions.py [hunk 28]
```diff
@@ -0,0 +1,75 @@
+import os
+import tempfile
+import pytest
+import json
+from flasksrc import create_app, db
+from flask_jwt_extended import create_access_token
+
+@pytest.fixture
+def client():
```
Comment (takakonishimura) [inline tests/test_sessions.py:9]: please also test the `put` method of `ToyNetSessionById` and then `get` again to ensure updates persist and are retrievable.

**Review by takakonishimura — CHANGES_REQUESTED** [review]
I will fix some of these print statements and the failing marshmallow unit tests.

File: flasksrc/session.py (hunk 15 above)
Comment (takakonishimura) [reply flasksrc/session.py:70]: remove 1 level of indentation

File: flasksrc/session.py (hunk 12 above)
Comment (takakonishimura) [reply flasksrc/session.py:122]: remove 1 level of indentation

File: flasksrc/session.py [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+    status = fields.Bool()
+    toynet_session_id = fields.Int()
+
+class ToyNetSessionByIdGetRequestSchema(Schema):
+    toynet_session_id = fields.Int()
+
+class ToyNetSessionByIdGetResponseSchema(Schema):
+    topo_id = fields.Int()
+    topology = fields.Str()
+    user_id = fields.Str()
```
[the remark this answers is not in the document]
Comment (blazebissar) [reply flasksrc/session.py:22]: Fixed

File: flasksrc/session.py [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+
+class ToyNetSessionByIdPutRequestSchema(Schema):
+    toynet_session_id = fields.Int()
```
[the remark this answers is not in the document]
Comment (blazebissar) [reply flasksrc/session.py]: Latest push resolved

File: flasksrc/session.py (hunk 1 above)
Comment (blazebissar) [reply flasksrc/session.py]: Resolved from latest push

File: flasksrc/session.py (hunk 4 above)
Comment (blazebissar) [reply flasksrc/session.py:79]: Field removed

File: flasksrc/session.py (hunk 3 above)
Comment (blazebissar) [reply flasksrc/session.py]: Removed

File: flasksrc/__init__.py (hunk 2 above)
Comment (blazebissar) [reply flasksrc/__init__.py:8]: Removed

Comment (blazebissar) [reply flasksrc/__init__.py:8]: Looks like the Hello World for the app/easy heartbeat/liveness check. We could delete in another refactor if needed.

File: flasksrc/session.py (hunk 3 above)
Comment (blazebissar) [reply flasksrc/session.py]: Removed.

File: flasksrc/session.py (hunk 7 above)
Comment (ajlacey) [reply flasksrc/session.py:91]: cleaned.

File: flasksrc/value.py (hunk 9 above)
Comment (ajlacey) [reply flasksrc/value.py]: already removed.

File: tests/test_values.py (hunk 10 above)
Comment (ajlacey) [reply tests/test_values.py]: already removed.

File: flasksrc/session.py (hunk 13 above)
Comment (ajlacey) [reply flasksrc/session.py:40]: updated.

File: flasksrc/session.py (hunk 5 above)
Comment (blazebissar) [reply flasksrc/session.py:94]: I attempted some naive SQL injection and it seems that the library handles this for us. Might be worth getting Arod to pen test for us.

File: flasksrc/user.py (hunk 6 above)
Comment (blazebissar) [reply flasksrc/user.py]: I think this is intended to give an additional indicator of success to the frontend on the user status? @takakonishimura ?

File: tests/test_sessions.py (hunk 17 above)
Comment (ajlacey) [reply tests/test_sessions.py:32]: updated

File: flasksrc/__init__.py (hunk 8 above)
Comment (blazebissar) [reply flasksrc/__init__.py:66]: Yes; confirmed in the DB schema that they are all integers.

File: tests/test_sessions.py (hunk 16 above)
Comment (ajlacey) [reply tests/test_sessions.py:25]: updated.

File: tests/test_sessions.py (hunk 21 above)
Comment (ajlacey) [reply tests/test_sessions.py:46]: removed.

File: tests/test_sessions.py (hunk 23 above)
Comment (ajlacey) [reply tests/test_sessions.py:54]: deleted

File: tests/test_sessions.py (hunk 19 above)
Comment (ajlacey) [reply tests/test_sessions.py:36]: removed.

File: tests/test_sessions.py (hunk 22 above)
Comment (ajlacey) [reply tests/test_sessions.py:51]: 'Loser'.lower

File: tests/test_sessions.py (hunk 24 above)
Comment (ajlacey) [reply tests/test_sessions.py:1]: replaced " with '

File: tests/test_sessions.py (hunk 25 above)
Comment (ajlacey) [reply tests/test_sessions.py:60]: removed.

File: tests/test_sessions.py (hunk 26 above)
Comment (ajlacey) [reply tests/test_sessions.py:62]: went with the latter.

File: tests/test_sessions.py (hunk 27 above)
Comment (ajlacey) [reply tests/test_sessions.py:72]: removed.

File: tests/test_sessions.py (hunk 11 above)
Comment (blazebissar) [reply tests/test_sessions.py:20]: Added

File: flasksrc/session.py (hunk 12 above)
Comment (blazebissar) [reply flasksrc/session.py:121]: Addressed

File: flasksrc/session.py (hunk 14 above)
Comment (blazebissar) [reply flasksrc/session.py:54]: Addressed

File: tests/test_sessions.py (hunk 18 above)
Comment (blazebissar) [reply tests/test_sessions.py:33]: Consolidated

File: tests/test_sessions.py (hunk 20 above)
Comment (blazebissar) [reply tests/test_sessions.py:43]: Done

File: tests/test_sessions.py (hunk 22 above)
Comment (blazebissar) [reply tests/test_sessions.py:51]: Done

File: tests/test_sessions.py (hunk 19 above)
Comment (blazebissar) [reply tests/test_sessions.py:36]: Removed

File: tests/test_sessions.py (hunk 23 above)
Comment (blazebissar) [reply tests/test_sessions.py:54]: Removed

File: tests/test_sessions.py (hunk 24 above)
Comment (blazebissar) [reply tests/test_sessions.py:1]: Replaced

File: tests/test_sessions.py (hunk 28 above)
Comment (ajlacey) [reply tests/test_sessions.py:9]: need to add a schema for ToyNetSessionByIdPutReq

**Review by takakonishimura — APPROVED** [review]
LGTM! Awesome work @ajlacey & @blazebissar !!!

**Pull request merged**